File: CS/Shared/Common.DataModel.EntityFramework/EntityFrameworkCRUDDataProvider.cs
using System;
using System.Data.Entity;
using System.Data.Entity.Infrastructure;
using System.Data.Entity.Validation;
using System.Linq;
using System.Linq.Expressions;
using System.Text;

namespace DevExpress.CRUD.DataModel.EntityFramework {
    public class EntityFrameworkCRUDDataProvider<TContext, TEntity, T, TKey> : EntityFrameworkDataProvider<TContext, TEntity, T>, ICRUDDataProvider<T>
        where TEntity : class
        where T : class
        where TContext : DbContext {

        readonly Func<T, TKey> getKey;
        readonly Func<TEntity, TKey> getEntityKey;
        readonly Action<T, TKey> setKey;
        readonly Action<T, TEntity> applyProperties;

        public EntityFrameworkCRUDDataProvider(
            Func<TContext> createContext, Func<TContext, DbSet<TEntity>> getDbSet, Expression<Func<TEntity, T>> getEnityExpression,
            Func<T, TKey> getKey, Func<TEntity, TKey> getEntityKey, Action<T, TKey> setKey, Action<T, TEntity> applyProperties)
            : base(createContext, getDbSet, getEnityExpression) {
            this.getKey = getKey;
            this.getEntityKey = getEntityKey;
            this.setKey = setKey;
            this.applyProperties = applyProperties;
        }

        void ICRUDDataProvider<T>.Delete(T obj) {
            using(var context = createContext()) {
                var entity = getDbSet(context).Find(getKey(obj));
                if(entity == null) {
                    throw new NotImplementedException("The modified row no longer exists in the database. Handle this case according to your requirements.");
                }
                getDbSet(context).Remove(entity);
                SaveChanges(context);
            }
        }

        void ICRUDDataProvider<T>.Create(T obj) {
            using(var context = createContext()) {
                var entity = getDbSet(context).Create();
                getDbSet(context).Add(entity);
                applyProperties(obj, entity);
                SaveChanges(context);
                setKey(obj, getEntityKey(entity));
            }
        }

        void ICRUDDataProvider<T>.Update(T obj) {
            using(var context = createContext()) {
                var entity = getDbSet(context).Find(getKey(obj));
                if(entity == null) {
                    throw new NotImplementedException("The modified row does not exist in a database anymore. Handle this case according to your requirements.");
                }
                applyProperties(obj, entity);
                SaveChanges(context);
            }
        }

        static void SaveChanges(TContext context) {
            try {
                context.SaveChanges();
            } catch(Exception e) {
                throw ConvertException(e);
            }
        }

        static DbException ConvertException(Exception e) {
            var entityValidationException = e as DbEntityValidationException;
            if(entityValidationException != null) {
                var stringBuilder = new StringBuilder();
                foreach(var validationResult in entityValidationException.EntityValidationErrors) {
                    foreach(var error in validationResult.ValidationErrors) {
                        if(stringBuilder.Length > 0)
                            stringBuilder.AppendLine();
                        stringBuilder.Append(error.PropertyName + ": " + error.ErrorMessage);
                    }
                }
                return new DbException(stringBuilder.ToString(), entityValidationException);
            }
            return new DbException("An error has occurred while updating the database.", entityValidationException);
        }
    }
    public class DbException : Exception {
        public DbException(string message, Exception innerException) 
            : base(message, innerException) { 
        } 
    }
}

File: VB/Shared/Common.DataModel.EntityFramework/EntityFrameworkCRUDDataProvider.vb
Imports System
Imports System.Data.Entity
Imports System.Data.Entity.Validation
Imports System.Linq.Expressions
Imports System.Text

Namespace DevExpress.CRUD.DataModel.EntityFramework

    Public Class EntityFrameworkCRUDDataProvider(Of TContext As DbContext, TEntity As Class, T As Class, TKey)
        Inherits EntityFrameworkDataProvider(Of TContext, TEntity, T)
        Implements ICRUDDataProvider(Of T)

        Private ReadOnly getKey As Func(Of T, TKey)

        Private ReadOnly getEntityKey As Func(Of TEntity, TKey)

        Private ReadOnly setKey As Action(Of T, TKey)

        Private ReadOnly applyProperties As Action(Of T, TEntity)

        Public Sub New(ByVal createContext As Func(Of TContext), ByVal getDbSet As Func(Of TContext, DbSet(Of TEntity)), ByVal getEnityExpression As Expression(Of Func(Of TEntity, T)), ByVal getKey As Func(Of T, TKey), ByVal getEntityKey As Func(Of TEntity, TKey), ByVal setKey As Action(Of T, TKey), ByVal applyProperties As Action(Of T, TEntity))
            MyBase.New(createContext, getDbSet, getEnityExpression)
            Me.getKey = getKey
            Me.getEntityKey = getEntityKey
            Me.setKey = setKey
            Me.applyProperties = applyProperties
        End Sub

        Private Sub Delete(ByVal obj As T) Implements ICRUDDataProvider(Of T).Delete
            Using context = createContext()
                Dim entity = getDbSet(context).Find(getKey(obj))
                If entity Is Nothing Then
                    Throw New NotImplementedException("The modified row no longer exists in the database. Handle this case according to your requirements.")
                End If

                getDbSet(context).Remove(entity)
                SaveChanges(context)
            End Using
        End Sub

        Private Sub Create(ByVal obj As T) Implements ICRUDDataProvider(Of T).Create
            Using context = createContext()
                Dim entity = getDbSet(context).Create()
                getDbSet(context).Add(entity)
                applyProperties(obj, entity)
                SaveChanges(context)
                setKey(obj, getEntityKey(entity))
            End Using
        End Sub

        Private Sub Update(ByVal obj As T) Implements ICRUDDataProvider(Of T).Update
            Using context = createContext()
                Dim entity = getDbSet(context).Find(getKey(obj))
                If entity Is Nothing Then
                    Throw New NotImplementedException("The modified row does not exist in a database anymore. Handle this case according to your requirements.")
                End If

                applyProperties(obj, entity)
                SaveChanges(context)
            End Using
        End Sub

        Private Shared Sub SaveChanges(ByVal context As TContext)
            Try
                context.SaveChanges()
            Catch e As Exception
                Throw ConvertException(e)
            End Try
        End Sub

        Private Shared Function ConvertException(ByVal e As Exception) As DbException
            Dim entityValidationException = TryCast(e, DbEntityValidationException)
            If entityValidationException IsNot Nothing Then
                Dim stringBuilder = New StringBuilder()
                For Each validationResult In entityValidationException.EntityValidationErrors
                    For Each [error] In validationResult.ValidationErrors
                        If stringBuilder.Length > 0 Then stringBuilder.AppendLine()
                        stringBuilder.Append([error].PropertyName & ": " & [error].ErrorMessage)
                    Next
                Next

                Return New DbException(stringBuilder.ToString(), entityValidationException)
            End If

            Return New DbException("An error has occurred while updating the database.", entityValidationException)
        End Function
    End Class

    Public Class DbException
        Inherits Exception

        Public Sub New(ByVal message As String, ByVal innerException As Exception)
            MyBase.New(message, innerException)
        End Sub
    End Class
End Namespace
